db/interface.go

```diff
@@ -1,62 +1,51 @@
-// it under the terms of the GNU Lesser General Public License as published by
-// the Free Software Foundation, either version 3 of the License, or
-// (at your option) any later version.
-//
-// The gossamer library is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-// GNU Lesser General Public License for more details.
-//
-// You should have received a copy of the GNU Lesser General Public License
-// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
-
-// This file is copied from ChainSafe/gossamer/lib/database.
-
 package db
 
-import "io"
+import (
+	"fmt"
+	"io"
+)
+
+// ErrKeysNotFound is used to indicate that a key does not exist in the db.
+var ErrKeyNotFound = fmt.Errorf("key not found")
 
-// Database wraps all database operations. All methods are safe for concurrent use.
+// Database wraps all database operations. All methods are safe for concurrent
+// use.
 type Database interface {
-	Reader
-	Writer
 	io.Closer
 
-	NewBatch() Batch
-	Path() string
-	NewIterator() Iterator
+	// ReadTx returns a ReadTx
+	ReadTx() ReadTx
+	// WriteTx returns a WriteTx
+	WriteTx() WriteTx
+	// Iterate iterates over the key-value pairs executing the given
+	// callback function for each pair. If a prefix is given, it will only
+	// iterate over the key-values with the prefix. The callback returns a
+	// boolean which is used to indicate to the iteration if to stop or to
+	// continue iterating. While the callback returns `true`, the iteration
+	// will keep continuing, and if the callback returned boolean is
+	// `false`, the iteration will stop.
+	Iterate(prefix []byte, callback func(key, value []byte) bool) error
 }
 
-// Batch is a write-only operation.
-type Batch interface {
-	Writer
-
-	ValueSize() int
-	Write() error
-	Reset()
-}
-
-// Iterator iterates over key/value pairs in ascending key order.
-// Must be released after use.
-type Iterator interface {
-	Next() bool
-	Key() []byte
-	Value() []byte
-	Release()
-}
-
-// Reader interface
-type Reader interface {
+type ReadTx interface {
+	// Get retreives the value for the given key. If the key does not
+	// exist, returns the error ErrKeyNotFound
 	Get(key []byte) ([]byte, error)
-	Has(key []byte) (bool, error)
+	// Discard discards the transaction. This method can be called always,
+	// even if previously the Tx has been Commited (for the WriteTx case).
+	// So it's a good practice to `defer tx.Discard()` just after creating
+	// the tx.
+	Discard()
 }
 
-// Writer interface
-type Writer interface {
-	Put(key []byte, value []byte) error
+type WriteTx interface {
+	ReadTx
+
+	// Set adds a key-value pair. If the key already exists, it's value is
+	// updated.
+	Set(key []byte, value []byte) error
+	// Del deletes a key and its value.
 	Del(key []byte) error
```

Comment:
I think I'm changing my mind on this name; Delete is barely any longer, and easier to read.

It's not like we are requiring all methods to be super short, e.g. Commit.